datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,211 @@
+import os
+from typing import Dict, List, Optional
+from datastore.datastore import DataStore
+from models.models import (
+    DocumentChunk,
+    DocumentChunkMetadata,
+    DocumentMetadataFilter,
+    DocumentChunkWithScore,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding,
+)
+import json
+import psycopg
+from psycopg import sql
+
+from services.date import to_unix_timestamp
+
+# Read environment variables for Postgres
+
+POSTGRES_HOST = os.environ.get("POSTGRES_HOST", "localhost")
+POSTGRES_PORT = int(os.environ.get("POSTGRES_PORT", 5433))
+POSTGRES_USERNAME = os.environ.get("POSTGRES_USERNAME", "postgres")
+POSTGRES_PASSWORD = os.environ.get("POSTGRES_PASSWORD", "")
+POSTGRES_DATABASE = os.environ.get("POSTGRES_DATABASE", "pgml_development")
+POSTGRES_TABLENAME = os.environ.get("POSTGRES_TABLENAME", "chatgpt_datastore")
+POSTGRES_SYNCHRONOUS_COMMIT = os.environ.get(
+    "POSTGRES_SYNCHRONOUS_COMMIT", "off"
+).lower()
+
+assert POSTGRES_SYNCHRONOUS_COMMIT in ["on", "off"]
+# OpenAI Ada Embeddings Dimension
+VECTOR_DIMENSION = 1536
+
+
+class PostgresDataStore(DataStore):
+    def __init__(self) -> None:
+        self.conn_params = {
+            "dbname": POSTGRES_DATABASE,
+            "user": POSTGRES_USERNAME,
+            "password": POSTGRES_PASSWORD,
+            "host": POSTGRES_HOST,
+            "port": POSTGRES_PORT,
+        }
+        conn = psycopg.connect(**self.conn_params)
+        # Insert the vector and text into the database
+        create_table = (
+            "CREATE TABLE IF NOT EXISTS %s (doc_id TEXT, chunk_id TEXT, text TEXT, embedding vector(%s), metadata JSONB)"
+            % (POSTGRES_TABLENAME, str(VECTOR_DIMENSION))
+        )
+        cur = conn.cursor()
+        cur.execute(create_table)
+        # Commit the transaction
+        conn.commit()
+        # Close the cursor and the connection
+        cur.close()
+        conn.close()
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
+        """
+        Takes in a dict from document id to list of document chunks and inserts them into the index.
+        Return a list of document ids.
+        """
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        # Initialize a list of ids to return
+        doc_ids: List[str] = []
+
+        # Set synchronous commit
+        cur.execute("SET synchronous_commit = %s" % POSTGRES_SYNCHRONOUS_COMMIT)
```

Comment:
Clients don't often change the commit policy in a connection like this, although I suppose it's fine, if they don't actually care? This seems like an abstraction leak that we should avoid exposing to high level clients in general though. Generally this is something a qualified DBA would weigh in on per use case.

Reply:
I will remove this. It seems unsafe too.